libimage/push.go

```diff
@@ -24,33 +30,93 @@ type PushOptions struct {
 //
 // Return storage.ErrImageUnknown if source could not be found in the local
 // containers storage.
+// Returns the bytes of the copied manifest when pushing a single tag,
+// which may be used for digest computation.
+// When pushing with AllTags=true then the returned []byte is always nil.
 func (r *Runtime) Push(ctx context.Context, source, destination string, options *PushOptions) ([]byte, error) {
 	if options == nil {
 		options = &PushOptions{}
 	}
 
-	// Look up the local image.  Note that we need to ignore the platform
-	// and push what the user specified (containers/podman/issues/10344).
-	image, resolvedSource, err := r.LookupImage(source, nil)
+	// Push the single image
+	if !options.AllTags {
+
+		// Look up the local image.  Note that we need to ignore the platform
+		// and push what the user specified (containers/podman/issues/10344).
+		image, resolvedSource, err := r.LookupImage(source, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure we have a proper destination, and parse it into an image
+		// reference for copying.
+		if destination == "" {
+			// Doing an ID check here is tempting but false positives (due
+			// to a short partial IDs) are more painful than false
+			// negatives.
+			destination = resolvedSource
+		}
+
+		return pushImage(ctx, image, destination, options, resolvedSource, r)
+	}
+
+	// Below handles the AllTags option, for which we have to build a list of
+	// all the local images that match the provided repository and then push them.
+	//
+	// For now, make sure a destination was not specified and get it from the source.
+	// This could change in the future, but that gets close to the Copy() functionality.
+	if len(destination) != 0 {
+		return nil, fmt.Errorf("`destination` should not be specified if using AllTags")
+	}
+
+	// Make sure the source repository does not have a tag
+	srcNamed, err := reference.ParseNormalizedNamed(source)
 	if err != nil {
 		return nil, err
 	}
+	if !reference.IsNameOnly(srcNamed) {
+		return nil, fmt.Errorf("can't push with AllTags if source tag is specified")
+	}
+
+	logrus.Debugf("Finding all images for source %s", srcNamed.Name())
+	listOptions := &ListImagesOptions{}
+	srcImages, _ := r.ListImages(ctx, []string{srcNamed.Name()}, listOptions)
```

Comment:
Does this actually work? AFAICS this is equivalent to
```go
srcImages = []*libimage.Image{r.LookupImage(srcNamed.Name()}
```
i.e. it always only finds one image. Am I missing something?

---

I increasingly think that the logic that chooses which images to push should have fairly robust unit tests (single image; single image with multiple tags; multiple tags pointing to different images; images with tags both in the specified repo and outside …). That’s hard to do with the current `Push` API, because it now doesn’t report anything about the returned images — so I guess the code to turn `source` into a list of `(*libimage.Image, reference.Named)`, or some similar output data, should be turned into a separate helper function with unit tests.

---

(Probably irrelevant: why is it OK to ignore errors here?)